feat(Accordion): add new component

carlafranca opened the pull request:
Jira ticket with the requirements
[DS-381](https://jira.equisoft.com/browse/DS-381)

All the Requirements are done except the item below that needs design
Add a placeholder or skeleton screen for the accordion panel when the content takes more than 2 seconds to load
All the Acceptance criteria in the ticket were tested.

Story created with information on how to use the component and all versions.

LarryMatte reviewed (changes requested):
A few CSS modifications are required 

To be as close to pixel perfect as we can, use the chrome extension [perfectpixels](https://chrome.google.com/webstore/detail/perfectpixel-by-welldonec/dkaagdgjmgdmbnecmcefdhjekcoceebi).

Plus, not quite sure about the animation. The padding makes it a bit weird _i.e._, the content moves as the padding increases. I'll try to find an example that could fix the issue.

JsGarneau reviewed (commented):
As far as code goes, it's clean 👌 

Now for the things talked during last meeting, here's my take;

- The wrapper is good, as I stated in my other comments, I'd just add more restriction to the children types so you don't have to make some awkward checks deeper in the component.

- The .toLowerCase() doesn't bother me, it's indeed an easy mistake to make when a user is implementing it to Javascript because he won't be forced to use proper typing (IDE might catch it but not enforce it). The thing to consider is that if we help users by correcting some mistakes, they might expect the same behaviour from other components and I don't like that.

- `disabledItemIds`, `defaultExpandedItemIds`; It's not a bad idea, I can see why in some case it would be preferred, what I don't like is that it kind of force you to add `id`, which wouldn't be necessary if we didn't have that and only used `Accordion` props to deal with these state.

carlafranca commented on packages/react/src/components/accordion/accordion-styles.ts:
```diff
@@ -0,0 +1,78 @@
+    &:hover,
+    &[aria-expanded='true'] {
+        background: ${theme.greys.white};
+        color: ${theme.greys['neutral-100']};
```
done, I removed the line &[aria-expanded='true'] but I need to keep the hover as the buttons has a different setup in the project then the one in the figma

carlafranca replied:
As per our conversation I'm keeping it with a TODO comment to be refactor/removed when we fix thematiation.

carlafranca commented on packages/react/src/components/accordion/accordion-styles.ts:
```diff
@@ -0,0 +1,78 @@
+        box-shadow: ${theme.tokens['focus-box-shadow-inset']};
+    }
+    > svg {
+        color: ${theme.greys['neutral-100']};
```
done

carlafranca commented on packages/storybook/stories/accordion.mdx:
```diff
@@ -0,0 +1,19 @@
+import * as AccordionStories  from './accordion.stories';
+
+
+# Accordion Comopnent
```
done

carlafranca commented on packages/react/src/components/accordion/accordion-styles.ts:
```diff
@@ -0,0 +1,78 @@
+    }
+    > svg {
+        color: ${theme.greys['neutral-100']};
+        height: 16px;
```
I got it rearranged, just bear in mind that to become pixel perfect I need to use margings 3px which is not part of the spacing setup (4, 8, 12 ....)

carlafranca commented on packages/react/src/components/accordion/accordion-styles.ts:
```diff
@@ -0,0 +1,78 @@
+    font-weight: var(--font-normal);
+    justify-content: start;
+    letter-spacing: 0.015rem;
+    line-height: 2;
```
done

carlafranca commented on packages/react/src/components/accordion/accordion-types.ts:
```diff
@@ -0,0 +1,29 @@
+import { Type, Tag } from '../heading/heading';
+
+export interface AccordionContainerProps {
+    children: ReactNode;
```
I can but it wont stop if someone add a div among the <Accordion>.
Please let me know if you know a way to do it as my typescript skills aren't pro yet.
Here is the implementation:
```
export interface AccordionContainerProps<T> {
    children: ReactElement<T> | ReactElement<T>[];
    /** Multipe or single panel open concurrently */
    mode?: 'single' | 'multi';
    /** Define which panel(s) should be open by default */
    defaultExpandedItemIds?: string[];
    /** Define which accordion should be disabled */
    disabledItemIds?: string[];
}
```
```
export const AccordionContainer: React.FC<AccordionContainerProps<AccordionProps>> = ({
    mode = 'single', children, defaultExpandedItemIds, disabledItemIds = [],
}) => {

const childrenArray: ReactElement<AccordionProps>[] = (
        React.Children.toArray(children) as ReactElement<AccordionProps>[])
        .map((child) => {
            const buttonRef = createRef<HTMLButtonElement>();
            const childProps = child.props as AccordionProps;
            const modifiedProps: AccordionProps = {
                ...childProps,
                buttonRef,
            };
            return React.cloneElement(child, modifiedProps);
        });
        ... rest of the code
```

carlafranca commented on packages/react/src/components/accordion/accordion-container.tsx:
```diff
@@ -0,0 +1,78 @@
+    };
+
+    const childrenArray: ReactElement<AccordionProps>[] = React.Children.toArray(children)
+        .filter((child): child is ReactElement<AccordionProps> => {
```
This doesn't stop someone to add a div
```
const childrenArray: ReactElement<AccordionProps>[] = (
        React.Children.toArray(children) as ReactElement<AccordionProps>[])
        .map((child) => {
            const buttonRef = createRef<HTMLButtonElement>();
            const childProps = child.props as AccordionProps;
            const modifiedProps: AccordionProps = {
                ...childProps,
                buttonRef,
            };
            return React.cloneElement(child, modifiedProps);
        });
```

The div below will be displayed but not as an Accordion element
```
export const Default: Story = () => (
    <AccordionContainer>
        <Accordion title="Panel Title 1" id="1">
            <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Dignissim viverra justo, nisl sit nascetur.</p>
        </Accordion>
        <Accordion title="Panel Title 2" id="2">
            <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Dignissim viverra justo, nisl sit nascetur.</p>
        </Accordion>
        <div>test</div>
        <Accordion title="Panel Title 3" id="3">
            <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Dignissim viverra justo, nisl sit nascetur.</p>
        </Accordion>
    </AccordionContainer>
);
```

meriouma commented on packages/react/src/components/accordion/index.tsx:
```diff
@@ -0,0 +1,2 @@
+export { Accordion } from './accordion';
+export { AccordionContainer } from './accordion-container';
```
Naming-wise, shouldn't the `AccordionContainer` be named `Accordion`, and the `Accordion` a `Collapsible`?

carlafranca replied:
do we have a name convention guide of some sort?

carlafranca replied:
I refactor the code to get the disabled and default expanded implemented per accordion item.
the code is a bit different now.

carlafranca replied:
ok renaming to Accordion/AccordionItem

carlafranca replied:
Name refactor is done. 
Now the Wrapper is <Accordion>
the children is <AccordionItem>

meriouma commented on packages/react/src/components/icon/icon.tsx:
```diff
@@ -51,6 +51,8 @@ import Organization from '../../icons/organization.svg';
 import Reorder from '../../icons/reorder.svg';
 import WarningFilled from '../../icons/warning-filled.svg';
 import Equisoft from '../../logos/logo-equisoft-ico.svg';
+import CarretDown from '../../icons/carret-down.svg';
+import CarretRight from '../../icons/carret-right.svg';
```
Should be `caret` and not `carret`

carlafranca replied:
done

carlafranca commented:
Thanks everyone for the feedback.
I refactor the code based on the input everyone has given.

1.I removed the option to set the disabled and defaultExpanded from the AccordionContainer.
Now it needs to be added per AccordionItem. (@savutsang, @JsGarneau) 

2. Also removed the id from AccordionItem, it is now generated within the accordionContainer component.
@JsGarneau: I'm adding an id in the accordionContainer to fix the issue with accessibility having duplicated id if we have more than one accordion in the same page.
I noticed this issue happens with tabs.

3. About the styles using perfectpixels, @LarryMatte you will see a lil difference, when I tried to match visually I was getting value as 3px or 11px in some cases so I rounded it to match 8ths.
Also fixed the animation.

4. @meriouma I renamed the icon.
Do you want me to change the <AccordionContainer>. to <Accordion> and 
<Accordion> to <AccordionItem>?
Please let me know which name do you prefer and I can change it.

Thanks

savutsang commented on packages/react/src/components/accordion/accordion-container.tsx:
```diff
@@ -0,0 +1,89 @@
+    const childrenArray: ReactElement<AccordionProps>[] = React.Children.toArray(children)
+        .filter(isAccordion)
+        .map((child: ReactElement<AccordionProps>, index) => {
+            const buttonRef = createRef<HTMLButtonElement>();
```
Some concerns with this ref.

- In function component, we should use `useRef` instead of createRef. 
- The ref should point to the same variable between refresh, while this code create a new ref every time.
- This ref itself has no use. You could forward a ref from the user to the target element, but if you create a ref yourself, it should have some code that use that ref (like some read/write on it).

savutsang replied:
Maybe just remove if not needed

carlafranca replied:
@savutsang I did a adjustment in the code to avoid regenerate ref between refresh. Let me explain some decisions I took to make sense on my refactor:
1. I'm using useMemo to ensures that the button refs are only created once and persist across renders, avoiding unnecessary re-creation.
2. With this implementation I can't use useRef as it's not allowed to use built-in hooks inside useMemo.

Note:
I saw the same implementation in tabs which is using useMemo with createRef in a functional component.
using createRef() within useMemo() seems to be valid and will create the refs once.

savutsang replied:
New code and clarification looks good to me.

savutsang commented on packages/react/src/components/accordion/accordion-container.tsx:
```diff
@@ -0,0 +1,89 @@
+        .filter(isAccordion)
+        .map((child: ReactElement<AccordionProps>, index) => {
+            const buttonRef = createRef<HTMLButtonElement>();
+            const childProps = child.props as AccordionProps;
```
This is unnecessary and redundant, also same in map.

- child.props is already resolved as AccordionProps because child is ReactElement<AccordionProps>.
- child is resolved as `ReactElement<AccordionProps>` because of the type guard isAccordion

You would still have all the typing like this.

```
React.Children.toArray(children)
  .filter(isAccordion)
  .map(child, index) => {
     child.props
  }
```

JsGarneau commented:
> Thanks everyone for the feedback. I refactor the code based on the input everyone has given.
> 
> 1.I removed the option to set the disabled and defaultExpanded from the AccordionContainer. Now it needs to be added per AccordionItem. (@savutsang, @JsGarneau)
> 
> 2. Also removed the id from AccordionItem, it is now generated within the accordionContainer component.
>    @JsGarneau: I'm adding an id in the accordionContainer to fix the issue with accessibility having duplicated id if we have more than one accordion in the same page.
>    I noticed this issue happens with tabs.
> 3. About the styles using perfectpixels, @LarryMatte you will see a pixel difference here. When I tried to match it visually I was having to use value as 3px or 11px in some cases so I rounded it to match 8ths.
>    Also fixed the animation.
> 4. @meriouma I renamed the icon.
>    Do you want me to change the AccordionContainer. to Accordion and
>    Accordion to AccordionItem?
>    Please let me know which name do you prefer and I can change it.
> 
> Thanks

Great!

I did some research about the typings of the children, I thought it would be as simple as to add the typings of the props on the `Array<ReactElement<AccordionItemProps>>`, but it's not. `React.Children.toArray()` seems to never return any typed `ReactChild` so you can safely ignore my earlier comments on the matter.

Good job!

carlafranca commented:
@LarryMatte I made some changes to the style to fix issues with animation.
Also added 
will-change: to inform the browser which properties will be animated, to help with optimization.
transform: translateZ(0): That triggers GPU acceleration to smooth the animation.

Please let me know if you find any issue.
thanks

LarryMatte reviewed (changes requested):
One minor adjustment, and then it should be just right.
Good work on the animation! It looks great!

carlafranca commented on packages/react/src/components/accordion/accordion-styles.ts:
```diff
@@ -0,0 +1,86 @@
+    justify-content: start;
+    letter-spacing: 0.015rem;
+    line-height: 1.5;
+    min-height: var(--spacing-5x); /* TODO change colors when updating thematization */
```
done

carlafranca commented on packages/react/src/components/accordion/accordion-styles.ts:
```diff
@@ -0,0 +1,86 @@
+    letter-spacing: 0.015rem;
+    line-height: 1.5;
+    min-height: var(--spacing-5x); /* TODO change colors when updating thematization */
+    padding: var(--spacing-1x); /* TODO change colors when updating thematization */
```
done

carlafranca commented on packages/react/src/components/accordion/accordion-styles.ts:
```diff
@@ -0,0 +1,86 @@
+    }
+    > svg {
+        height: 1rem;
+        margin: 3px var(--spacing-1halfx) 0 3px; /* TODO change colors when updating thematization */
```
done

carlafranca commented on packages/react/src/components/accordion/accordion-styles.ts:
```diff
@@ -0,0 +1,86 @@
+    width: 100%;
+    &:hover, /* TODO change colors when updating thematization */
+    &[aria-expanded='true'] {/* TODO change colors when updating thematization */
+        background: ${theme.greys.white};
```
background colour updated

LarryMatte commented on packages/react/src/components/accordion/accordion-styles.ts:
```diff
@@ -0,0 +1,86 @@
+    text-transform: none;
+    transition: border-radius 0.2s ease;
+    width: 100%;
+    &:hover, /* TODO change colors when updating thematization */
```
The hover and expanded states of the button currently feature different background colors in Figma:

Hover state: Grey background.
Expanded state: White background.
Expanded while hovered: Grey background.

However, in your PR, there's an issue where the expanded variant have a grey background even when the button is not hovered.

Hover:
https://www.figma.com/file/s0iqD0TbWNNt6anLNGS9bQ/Documentation%2C-Notion?type=design&node-id=26969-189929&mode=dev

LarryMatte replied:
Also, the border should be darker when the button is in the hover state.
`#1B1C1E`

carlafranca replied:
updated the colors on hover and expanded states.
Also updated the token name for #1B1C1E as per my last huddle with Mark-Andre as the neutral-100 will be the black token.

meriouma commented on packages/react/src/components/accordion/accordion-styles.ts:
```diff
@@ -0,0 +1,96 @@
+    &.expanded {
+        border: 1px solid ${theme.greys['neutral-15']};
+        border-top-width: 0;
+        max-height: 500px;
```
This could be problematic if the content has more than 500px.

carlafranca replied:
I can make it 99999px as it doesnt affect and I doubt anyone will have anything with that size right?
but I need to implement it due to height is not useful for the animation.
If you have any suggestion here I would appreciate.

meriouma commented on packages/react/src/components/accordion/accordion-styles.ts:
```diff
@@ -0,0 +1,96 @@
+    justify-content: start;
+    letter-spacing: 0.015rem;
+    line-height: 1.5;
+    min-height: var(--spacing-5x); /* TODO change space when updating thematization */
+    padding: var(--spacing-1x); /* TODO change space when updating thematization */
```
What is the TODO here?

carlafranca replied:
Done, removed the TODO comment

meriouma commented on packages/react/src/components/accordion/accordion-styles.ts:
```diff
@@ -0,0 +1,96 @@
+    transition: border-radius 0.2s ease;
+    width: 100%;
+
+    &[aria-expanded='true'] {/* TODO change colors when updating thematization */
```
And here?

meriouma commented on packages/react/src/components/accordion/accordion-styles.ts:
```diff
@@ -0,0 +1,96 @@
+        width: 1rem;
+    }
+
+    &:disabled { /* TODO change colors when updating thematization */
```
Lot of TODOs in this file. Is there really something pending for those?

carlafranca replied:
Thematization is the block.
unless you prefer not to have the  TODOs I'm following the other components update and being very specific

carlafranca replied:
Done, removed all TODO commments on style file

meriouma commented on packages/react/src/components/accordion/accordion-styles.ts:
```diff
@@ -0,0 +1,96 @@
+    &[aria-expanded='true'] {/* TODO change colors when updating thematization */
+        background: ${theme.greys.white};
+        color: ${theme.greys['neutral-90']};
+        transition: border-radius 0.2s ease;
```
The radius animation is a bit weird when collapsing, but I'm not sure if there's something we can do.

carlafranca replied:
added delay to improve transition on collapsing. hope that will look better.

meriouma commented on packages/react/src/components/accordion/accordion-item.tsx:
```diff
@@ -0,0 +1,77 @@
+import { Icon } from '../icon/icon';
+import { Heading } from '../heading/heading';
+
+const AccordionSection = styled.section<AccordionBodyProps>`
```
I think it would be easier to follow the code if the `AccordionBodyProps` interface was declared closer to it's usage.

carlafranca replied:
if I understood what you said, this comment is being kind of relate to separate style and interface.
I see that you are not fan of having kind of a separation of concerns, removing styles and types from the component itself. 
is this causing any issue? I know it is not a good practice but maybe I'm missing something here.
Could you let me know if this is causing some issues?

meriouma commented on packages/react/src/components/accordion/accordion-types.ts:
```diff
@@ -0,0 +1,31 @@
+}
+
+export interface AccordionItemProps {
+    /** Title Label */
```
Not sure this comment adds something

carlafranca replied:
this comment is going into storybook when displaying the props there.
I believe they are removed when compile the project to final use right?

carlafranca replied:
removed these comments that was used on story book

meriouma commented on packages/react/src/components/accordion/accordion-types.ts:
```diff
@@ -0,0 +1,31 @@
+    title: string;
+    /** Unique Id on each accordion per group */
+    id?: string;
+    /** This is relate to styling */
```
This one too

carlafranca replied:
removed the comments that was being used on storybook

meriouma commented on packages/react/src/components/accordion/accordion-types.ts:
```diff
@@ -0,0 +1,31 @@
+    id?: string;
+    /** This is relate to styling */
+    type?: Type;
+    /** Choose the right tag for page outline */
```
And this one

carlafranca replied:
this comment is going into storybook when displaying the props there.
I believe they are removed when compile the project to final use right?

carlafranca replied:
done, put this description in stories.

meriouma commented on packages/react/src/components/accordion/accordion-types.ts:
```diff
@@ -0,0 +1,31 @@
+    type?: Type;
+    /** Choose the right tag for page outline */
+    tag?: Tag;
+    isExpanded?: boolean;
```
```suggestion
    expanded?: boolean;
```

carlafranca replied:
done

meriouma commented on packages/react/src/components/accordion/accordion-types.ts:
```diff
@@ -0,0 +1,31 @@
+}
+
+export interface AccordionBodyProps {
+    isExpanded?: boolean;
```
```suggestion
    expanded?: boolean;
```

carlafranca replied:
done

meriouma commented on packages/react/src/components/accordion/accordion-item.test.tsx:
```diff
@@ -0,0 +1,46 @@
+            </AccordionItem>,
+        );
+
+        const accordionItem = wrapper.find('.accordion-content').at(0);
+        const accordionButton = wrapper.find('.accordion-button').at(0);
```
Use `data-testid` instead of classes to find elements in your tests.

carlafranca replied:
I will see how to use data-tesetid.... do we have a documentation for this of I will need to dig in the files to see how it is used?

carlafranca replied:
using the elements instead of data-testid or class

meriouma reviewed (commented):
First round of review. I'll try to finish tomorrow.

carlafranca commented on packages/react/src/components/accordion/accordion-item.tsx:
```diff
@@ -0,0 +1,77 @@
+import { Icon } from '../icon/icon';
+import { Heading } from '../heading/heading';
+
+const AccordionSection = styled.section<AccordionBodyProps>`
+    ${({ theme }) => AccordionSectionStyled(theme)};
+`;
+
+const AccordionBody = styled.div<AccordionBodyProps>`
+    ${({ theme }) => AccordionBodyStyled(theme)};
+`;
+
+const StyledHeading = styled(Heading)`
+    ${HeadingStyled};
+`;
+
+const StyledButton = styled(Button)<{ isExpanded: boolean }>`
+    ${({ isExpanded, theme }) => ButtonStyled(isExpanded, theme)};
+`;
```
I will have a look at this, but still think the style should be separated from the component declaration as we have some components with several files and to update the style is a bit messy.

carlafranca replied:
Please let me know how you see this done in a better way and I will implement it. thanks for the feedback.

carlafranca commented on packages/react/src/components/accordion/accordion-item.tsx:
```diff
@@ -0,0 +1,77 @@
+
+    return (
+        <>
+            <StyledHeading className="accordion-header" type={type} tag={tag} noMargin>
```
this is a option I added so the users can add extra style if they need it without having to make a mess on the component default style. I see that being very useful in the pass.
Any reason why you don't like the idea?

carlafranca replied:
left the className as per your request.

carlafranca commented on packages/react/src/components/accordion/accordion-item.tsx:
```diff
@@ -0,0 +1,77 @@
+            <StyledHeading className="accordion-header" type={type} tag={tag} noMargin>
+                <StyledButton
+                    id={headerId}
+                    className="accordion-button"
```
this is a option I added so the users can add extra style if they need it without having to make a mess on the component default style. I see that being very useful in the pass.
Any reason why you don't like the idea?

carlafranca replied:
I removed the className flexibility as per your request.

carlafranca commented on packages/react/src/components/accordion/accordion-item.tsx:
```diff
@@ -0,0 +1,77 @@
+                aria-disabled={disabled}
+            >
+                <AccordionBody
+                    className="accordion-body"
```
this is a option I added so the users can add extra style if they need it without having to make a mess on the component default style. I see that being very useful in the pass.
Any reason why you don't like the idea?

carlafranca replied:
I removed all the className flexibility added previously as per your request.

carlafranca commented on packages/react/src/components/accordion/accordion-item.tsx:
```diff
@@ -0,0 +1,77 @@
+                </StyledButton>
+            </StyledHeading>
+            <AccordionSection
+                className={`accordion-content ${isExpanded ? 'expanded' : ''}`}
```
this is a option I added so the users can add extra style if they need it without having to make a mess on the component default style. I see that being very useful in the pass.
Any reason why you don't like the idea?

carlafranca commented on packages/react/src/components/accordion/accordion-item.tsx:
```diff
@@ -0,0 +1,77 @@
+    id,
+    type = 'medium',
+    tag = 'h3',
+    isExpanded = false,
```
done

carlafranca commented on packages/react/src/components/accordion/accordion-item.tsx:
```diff
@@ -0,0 +1,77 @@
+                    label={title}
+                    aria-expanded={isExpanded}
+                    aria-controls={panelId}
+                    onClick={() => onToggle()}
```
done

carlafranca commented on packages/react/src/components/accordion/accordion-styles.ts:
```diff
@@ -0,0 +1,96 @@
+    justify-content: flex-start;
+`;
+
+export const AccordionSectionStyled: AccordionSectionFunction = (theme) => css`
```
done

carlafranca commented on packages/react/src/components/accordion/accordion-styles.ts:
```diff
@@ -0,0 +1,96 @@
+    background: ${theme.greys['neutral-02']};
+    border: 0 solid ${theme.greys['neutral-15']};
+    border-radius: 0 0 var(--border-radius-2x) var(--border-radius-2x);
+    border-top-width: 0;
```
I removed the reference to the new colours name but we have a new colour that doesn't exist in the greys and we need to use this new grey in the accordion.
How would you like me to address it? using the new naming or not we will still need the colour #1B1C1E to be used. 

This is the new colour:
'neutral-90': '#1B1C1E',

This is how I implement the greys.ts:
```
export const greys: Theme['greys'] = {
    white: '#FFFFFF',
    'neutral-00': '#FFFFFF',
    'neutral-02': '#FAFAFA',
    'neutral-05': '#F1F2F2',
    'neutral-15': '#DBDEE1',
    'neutral-30': '#B7BBC2',
    'neutral-50': '#878F9A',
    'neutral-65': '#60666E',
    'neutral-90': '#1B1C1E',
    'colored-white': '#FAFAFA',
    'light-grey': '#F1F2F2',
    grey: '#DBDEE1',
    'mid-grey': '#B7BBC2',
    'dark-grey': '#60666E',
    black: '#000000',
};
```

This is how I'm using
${theme.greys['neutral-90']};

Note:
All the other references I added to the new naming I removed except the extra non-existent colour.

carlafranca commented on packages/react/src/components/accordion/accordion-styles.ts:
```diff
@@ -0,0 +1,96 @@
+    max-height: 0;
+    overflow: hidden;
+    transition: max-height 0.5s ease, border 0.5s ease;
+    &.expanded {
```
done

LarryMatte commented on packages/react/src/components/accordion/accordion-item.tsx:
```diff
@@ -0,0 +1,166 @@
+                id={panelId}
+                aria-labelledby={headerId}
+                aria-expanded={isExpanded}
+                role="region"
```
You probably don't need the `role="region"` since the HTML tag is a `<section>`.

> Using the `<section>` element will automatically communicate that a section has a role of region if it is given an accessible name. If at all possible, prefer using `<section>` instead.
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/region_role

carlafranca replied:
you are right.

LarryMatte commented on packages/react/src/components/accordion/accordion-item.tsx:
```diff
@@ -0,0 +1,166 @@
+    font-weight: var(--font-normal);
+    letter-spacing: 0.015rem;
+    line-height: 1.7;
+    padding: var(--spacing-2x) var(--spacing-3x) var(--spacing-3x) var(--spacing-5x);
```
There's a subtle difference in the `padding` due to the `margin` of the paragraph. 
![Screenshot 2023-09-14 at 11 02 33 AM](https://github.com/kronostechnologies/design-elements/assets/3722754/cce00088-d9d6-443f-8980-c4e0a148485b)

You might consider a solution like this:

```
> p:first-child {
        margin-top: 0;
    }
```
Feel free to explore other options as well. ;)

carlafranca replied:
I guess this will aways be an issue because users can add text and/or html elements.

savutsang replied:
I believe it's more the responsibility of the user to eliminate these margins. As @carlafranca said, there is just an infinite possibility, you have many elements (H2, UL, TABLE, P, etc), and they can even be nested.
```
<div>
    <div>
        <p>
    <div>
</div>
```

For the tests, I think we should simply use plain text (or to the limit just div).

carlafranca replied:
fixed as per our conversation.

LarryMatte reviewed (changes requested):
One last thing 😉: When there's a disabled accordion between two active ones, the use of the arrow up or down keys doesn't work.

For example, if the 1st accordion item is enabled and focused, the 2nd is disabled, and the 3rd is enabled, using the arrow down key to navigate between the 1st item and the 3rd one won't work as expected.

carlafranca commented:
Arrow navigation with disabled items fixed

meriouma commented on packages/react/src/components/accordion/accordion-item.tsx:
```diff
@@ -0,0 +1,173 @@
+    transition-timing-function: ease;
+    width: 100%;
+
+    &.expanded {
```
I think you could rely on `data-expanded` instead of adding a CSS class to the component.

carlafranca replied:
done

meriouma commented on packages/react/src/components/accordion/accordion-item.tsx:
```diff
@@ -0,0 +1,173 @@
+    max-height: 0;
+    overflow: hidden;
+    transition: max-height 0.5s ease, border-width 0.5s ease;
+    &.expanded {
```
Here too

carlafranca replied:
done

meriouma reviewed (changes requested):
I think there's an issue with the API of the AccordionItem.

LarryMatte commented on packages/react/src/components/accordion/accordion-item.tsx:
```diff
@@ -0,0 +1,173 @@
+    overflow: hidden;
+    transition: max-height 0.5s ease, border-width 0.5s ease;
+    &.expanded {
+        border-width: 1px;
```
`border-width: 0 1px 1px 1px;` so you wouldn't need the `border-top-width: 0;` on the next line.

carlafranca replied:
done

LarryMatte reviewed (dismissed):
LGTM as far as I'm concerned _i.e._, in terms of a11y, styling and behavior.

carlafranca commented on packages/react/src/components/accordion/accordion.tsx:
```diff
@@ -0,0 +1,130 @@
+    expanded?: boolean | undefined;
+    disabled?: boolean | undefined;
+    onToggle?: () => void;
+    onKeyDown?: ((event: React.KeyboardEvent<HTMLButtonElement>) => void) | undefined;
```
done

carlafranca commented on packages/react/src/components/accordion/accordion.tsx:
```diff
@@ -0,0 +1,130 @@
+`;
+
+export const Accordion: React.FC<AccordionProps> = ({
+    id, mode = 'single', children,
```
done

carlafranca commented on packages/react/src/components/accordion/accordion.tsx:
```diff
@@ -0,0 +1,130 @@
+    );
+
+    const isAccordion = (child: React.ReactNode): child is React.ReactElement<AccordionItemProps> => (
+        React.isValidElement<AccordionItemProps>(child) && child.type === AccordionItem);
```
done

carlafranca commented on packages/react/src/components/accordion/accordion-item.tsx:
```diff
@@ -0,0 +1,173 @@
+    title: string;
+    id?: string;
+    type?: Type | undefined;
+    tag?: Tag | undefined;
```
done

carlafranca commented on packages/react/src/components/accordion/accordion.tsx:
```diff
@@ -0,0 +1,130 @@
+    const isAccordion = (child: React.ReactNode): child is React.ReactElement<AccordionItemProps> => (
+        React.isValidElement<AccordionItemProps>(child) && child.type === AccordionItem);
+
+    const filteredChildren = React.Children.toArray(children).filter(isAccordion);
```
done

carlafranca commented on packages/react/src/components/accordion/accordion.tsx:
```diff
@@ -0,0 +1,130 @@
+import { AccordionItem } from './accordion-item';
+
+interface AccordionProps {
+    id: string;
```
done

carlafranca commented on packages/react/src/components/accordion/accordion.tsx:
```diff
@@ -0,0 +1,130 @@
+            return React.cloneElement(child, accordionProps);
+        }),
+        [filteredChildren, generateId, handleToggle],
+    );
```
@meriouma I was checking about making this refactor and I though about using context. What do you think?
